### PR TITLE
Add nginx-ldapauth-proxy

### PR DESCRIPTION
**Pull request by rendhalver**
We created this as an authentication layer on top of our ELK stack.
We thought it might be useful for other people as well.

*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

### CODE REVIEW

**Review by jar361 — APPROVED** [review]

File: incubator/nginx-ldapauth-proxy/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline incubator/nginx-ldapauth-proxy/templates/deployment.yaml]: `apps/v1beta2`

File: incubator/nginx-ldapauth-proxy/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+kind: Deployment
+metadata:
+  name: {{ template "nginx-ldapauth-proxy.fullname" . }}
+  labels:
+    app: {{ template "nginx-ldapauth-proxy.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
```
Comment (unguiculus) [inline incubator/nginx-ldapauth-proxy/templates/deployment.yaml]: Add `spec.selector`

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/nginx-ldapauth-proxy/templates/deployment.yaml (hunk 2 above)
Comment (rendhalver) [reply incubator/nginx-ldapauth-proxy/templates/deployment.yaml]: I think my latest commit addresses that.
Is that what you meant?

Comment (unguiculus) [reply incubator/nginx-ldapauth-proxy/templates/deployment.yaml]: The selector is not correct. You need to fix the `app` label and add the `release` label.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Review by unguiculus — CHANGES_REQUESTED** [review]
Please compare the `_helpers.tpl` you get with `helm create` using Helm 2.8+ and update accordingly. If you like you can move the chart to stable in this PR.

File: incubator/nginx-ldapauth-proxy/templates/configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "nginx-ldapauth-proxy.fullname" . }}
+  labels:
+    app: {{ template "nginx-ldapauth-proxy.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
```
Comment (unguiculus) [inline incubator/nginx-ldapauth-proxy/templates/configmap.yaml]: `chart: {{ template "nginx-ldapauth-proxy.chart" . }}`

File: incubator/nginx-ldapauth-proxy/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  name: {{ template "nginx-ldapauth-proxy.fullname" . }}
+  labels:
+    app: {{ template "nginx-ldapauth-proxy.name" . }}
+    chart: {{ template "nginx-ldapauth-proxy.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  selector:
+    matchLabels:
+      app: {{ template "nginx-ldapauth-proxy.fullname" . }}
```
Comment (unguiculus) [inline incubator/nginx-ldapauth-proxy/templates/deployment.yaml]: `app: {{ template "nginx-ldapauth-proxy.name" . }}`

File: incubator/nginx-ldapauth-proxy/templates/ingress.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+{{- if .Values.ingress.enabled -}}
+{{- $serviceName := include "nginx-ldapauth-proxy.fullname" . -}}
+{{- $servicePort := .Values.service.externalPort -}}
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "nginx-ldapauth-proxy.fullname" . }}
+  labels:
+    app: {{ template "nginx-ldapauth-proxy.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
```
Comment (unguiculus) [inline incubator/nginx-ldapauth-proxy/templates/ingress.yaml]: `chart: {{ template "nginx-ldapauth-proxy.chart" . }}`

File: incubator/nginx-ldapauth-proxy/templates/service.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "nginx-ldapauth-proxy.fullname" . }}
+  labels:
+    app: {{ template "nginx-ldapauth-proxy.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
```
Comment (unguiculus) [inline incubator/nginx-ldapauth-proxy/templates/service.yaml]: `chart: {{ template "nginx-ldapauth-proxy.chart" . }}`

File: incubator/nginx-ldapauth-proxy/templates/secrets.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ template "nginx-ldapauth-proxy.fullname" . }}
+  labels:
+    app: {{ template "nginx-ldapauth-proxy.name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
```
Comment (unguiculus) [inline incubator/nginx-ldapauth-proxy/templates/secrets.yaml]: `chart: {{ template "nginx-ldapauth-proxy.chart" . }}`

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by rendhalver** [comment]
@unguiculus I think I have addressed all your concerns.

**Review by unguiculus — APPROVED** [review]

**Comment by rendhalver** [comment]
This is failing because there is no service for it to connect to.
This is the error the pod gives me when I run it locally.
```
13:44 $ kubectl --context=docker-for-desktop --namespace=testing logs nginx-proxy-nginx-ldapauth-proxy-84cd45586d-b24g7
2018/03/02 18:45:12 [emerg] 1#1: host not found in upstream "backend-service.default.svc.cluster.local:8080" in /etc/nginx/nginx.conf:11
nginx: [emerg] host not found in upstream "backend-service.default.svc.cluster.local:8080" in /etc/nginx/nginx.conf:11
```
I guess I could cheat and connect it to a service that always exists but I am not sure what that would be maybe the kubernetes service?

I can verify that this does work as-is once you point it to a valid service and give it LDAP credentials.
We are using it for all of our elastic-stack clusters and it works well.

**Comment by rendhalver** [comment]
Is there a way to tell the tests to spin up another pod so I can use a service for testing this?
I can't see any other way of getting the tests to pass.
Failing that is there a service that will always exist in Kube that I can point this at to test?

**Comment by rjkernick** [comment]
You can write the test that helm will use. I don't see why you couldn't do a multi-container pod, one of which being a container for nginx to proxy. https://github.com/kubernetes/helm/blob/master/docs/chart_tests.md

**Comment by rendhalver** [comment]
Ahh that's handy! Thanks @rjkernick!

**Comment by unguiculus** [comment]
@rendhalver Any change you can continue with this?

**Comment by rendhalver** [comment]
The container I am using seems to fail if it can't get to the specified LDAP server so I need a way to setup an LDAP container to test the auth with.
I believe I can add some code to make that happen I just need to requisition some time.
I will see what I can do this weekend.

**Comment by rendhalver** [comment]
@unguiculus I finally got this passing tests.
I basically disable the ldap parts when there is no ldap host set.

File: stable/nginx-ldapauth-proxy/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      app: {{ template "nginx-ldapauth-proxy.name" . }}
+      release: {{ .Release.Name }}
+  replicas: {{ .Values.replicaCount }}
+  template:
+    metadata:
+      labels:
+        app: {{ template "nginx-ldapauth-proxy.name" . }}
+        release: {{ .Release.Name }}
+      annotations:
+        checksum/config: {{ include (print $.Template.BasePath "/configmap.yaml") . | sha256sum }}
```
Comment (unguiculus) [inline stable/nginx-ldapauth-proxy/templates/deployment.yaml]: The `$` is not necessary here.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/nginx-ldapauth-proxy/templates/deployment.yaml (hunk 8 above)
Comment (rendhalver) [reply stable/nginx-ldapauth-proxy/templates/deployment.yaml]: Oops.
Thanks. I will fix that.

**Review by unguiculus — APPROVED** [review]

**Pull request merged**